drivers/CAN.cpp

```diff
@@ -115,9 +115,14 @@ int CAN::filter(unsigned int id, unsigned int mask, CANFormat format, int handle
 void CAN::attach(Callback<void()> func, IrqType type) {
     lock();
     if (func) {
+        // lock deep sleep only the first time
+        if (_irq[(CanIrqType)type] == callback(donothing)) {
+            sleep_manager_lock_deep_sleep();
+        }
         _irq[(CanIrqType)type] = func;
         can_irq_set(&_can, (CanIrqType)type, 1);
     } else {
+        sleep_manager_unlock_deep_sleep();
```

Comment:
Same problem could occur with the else case - it is valid to call attach with null multiple times in a row. Maybe as a general strategy, each interrupt type could be locked when `donothing` is replaced, and unlocked when `donothing` is set. A similar change should also be made for SerialBase.